### ISSUE TITLE
Area2D's signals *_entered and *_exited do not fire predictably during movement dependent on speed and size.

### ISSUE THREAD

**Issue by julsams** [godotengine/godot#43743]
**Godot version:**
v3.2.3.stable.official

**OS/device including version:**
Windows 10 Home 64-bit (10.0 Build 19041)

**Issue description:**
Area2D's signals body_entered, body_exitedm, area_entered and area_exited do not fire predictably dependent on the size and speed of the objects involved.

**Steps to reproduce:**
1. Create an Area2D segment that splits the viewing area.
2. Create two RigidBody2D objects with a small circle as a child CollisionShape2D.
3. Set their linear_velocity such that the two RigidBody2D objects move toward and pass through the segment.
4. Increase the speed of one of them sufficiently and the signal will not fire anymore.
5. Slowing down the object or increasing its size will cause the signal to fire again.

**Minimal reproduction project:**
[enter_exit_non_detection.zip](https://github.com/godotengine/godot/files/5577361/enter_exit_non_detection.zip)

**Comment by Calinou** [comment]
This is yet another physics tunneling issue. Try increasing Physics Fps in the Project Settings.

**Comment by pouleyKetchoupp** [comment]
Confirmed on 3.2.3 and 3.2.4 beta 5.

Increasing the physics FPS a lot can fix the problem, but it can slow the game down by making physics run too many times for one game frame.
Setting the rigid body to Continuous collision detection should help but it doesn't in this case, that needs to be investigated.

**Comment by djrain** [comment]
The same issue exists for a KinematicBody2D moving quickly, is this expected?

I'm not familiar with the inner workings, but it seems like KinematicBody2D uses some kind of CCD by default. So in theory, shouldn't it be able to reliably trigger any Area2D it passes through?

**Comment by pouleyKetchoupp** [comment]
@djrain It's probably expected, I don't think KinematicBody2D does proper CCD. But it should be added at least as an option.

**Comment by AttackButton** [comment]
Any chance to implement the https://github.com/godotengine/godot/pull/69934 and https://github.com/godotengine/godot/pull/70160 for Area2D and Area3D? At least as an option on the editor (for performance reasons).

The problem at the moment is that area nodes only trigger the `body_entered` or `area_entered` signals 6 or 7 frames late.

I don't think changing the physics ticks is the best solution for this.

**Comment by Torguen** [comment]
Shouldn't this have a milestone 3.6?
I don't know, I see it as an important bug.

**Comment by Torguen** [comment]
Still in 3.6 beta 4

**Comment by laurajayoconnor** [comment]
> This is yet another physics tunneling issue. Try increasing Physics Fps in the Project Settings.

This solved the issue for me by dramatically upping physics ticks per second to 1,000 and max physics steps per frame to 100. I'm not sure how these settings will affect slower machines but it has no obvious downside on mine.

**Comment by Torguen** [comment]
I hope someone can solve it, I have it pending in my project and I have it marked as a pending bug although I applied a remedy I hope someone solves this.

https://discord.com/channels/212250894228652034/1114293117336690778

**Comment by Torguen** [comment]
Forgive my insistence but... I cry out to the gods so that some charitable soul can solve this.
I hope that Godot 3 support is not ended with the existence of this bug!

**Comment by Torguen** [comment]
This problem worsens in v4.5.dev1.official [97241ffea]. Now none of the bodies are detected.

In Godot v3.2.3.stable.official, one of the bodies is detected, the error occurs in the other body.

But in Godot v4.5.dev1.official [97241ffea], none of the bodies are detected.

Therefore, the new affected versions are: The upcoming v3.7.dev.custom_build [5155fe5a4] v4.5.dev1.official [97241ffea]

**Comment by starry-abyss** [comment]
> Forgive my insistence but... I cry out to the gods so that some charitable soul can solve this.
> 
> I hope that Godot 3 support is not ended with the existence of this bug!

Could you ask for the list everything overlapping with an area every frame as a workaround, so to continue developing the game instead of waiting?

**Comment by Torguen** [comment]
> > Forgive my insistence but... I cry out to the gods so that some charitable soul can solve this.
> > I hope that Godot 3 support is not ended with the existence of this bug!
> 
> Could you ask for the list everything overlapping with an area every frame (instead of signals) as a workaround, so to continue developing the game instead of waiting?

I’m already applying a workaround, but the best option is always for the engine to work as it should, I think we all agree on that.

It didn’t work for me to check with a list, the result obtained is the same as in the test, the fast body is not registered and doesn’t enter the list. I think you were referring to this, if not, could you kindly clarify what you meant?


With this code, 'now' is only written once, the fast body is not detected and is not registered in the list returned by 'get_overlapping_bodies':

func _physics_process(delta: float) -> void:
	
	if $Area2D.get_overlapping_bodies().size() > 0:
		print( "now" )

**Comment by starry-abyss** [comment]
@Torguen Yes, that's the method I was referring to. Sorry, I've missed the point about the impact of object speed on the issue. For this case the game engines I've worked with have an option called continuous collision detection (CCD). With it on, they should register the overlaps even between ticks, otherwise they shouldn't. I remember there is such an option in Godot too but didn't test it myself. So the option doesn't work, correct?
